src/storage/src/render/upsert.rs

```diff
@@ -590,131 +775,106 @@ where
                 input.next().await
             }
         } {
-            match event {
-                AsyncEvent::Data(_cap, mut data) => {
-                    stage_input(&mut stash, &mut data, &input_upper, &resume_upper, upsert_config.shrink_upsert_unused_buffers_by_ratio);
-                }
-                AsyncEvent::Progress(upper) => {
-                    // Ignore progress updates before the `resume_upper`, which is our initial
-                    // capability post-snapshotting.
-                    if PartialOrder::less_than(&upper, &resume_upper) {
-                        continue;
-                    }
-                    stash.sort_unstable();
-
-                    // Find the prefix that we can emit
-                    let idx = stash.partition_point(|(ts, _, _, _)| !upper.less_equal(ts));
-
-                    // Read the previous values _per key_ out of `state`, recording it
-                    // along with the value with the _latest timestamp for that key_.
-                    commands_state.clear();
-                    for (_, key, _, _) in stash.iter().take(idx) {
-                        commands_state.entry(*key).or_default();
-                    }
-
-                    // These iterators iterate in the same order because `commands_state`
-                    // is an `IndexMap`.
-                    multi_get_scratch.clear();
-                    multi_get_scratch.extend(commands_state.iter().map(|(k, _)| *k));
-                    match state
-                        .multi_get(multi_get_scratch.drain(..), commands_state.values_mut())
-                        .await
-                    {
-                        Ok(_) => {}
-                        Err(e) => {
-                            process_upsert_state_error::<G>(
-                                "Failed to fetch records from state".to_string(),
-                                e,
-                                &mut health_output,
-                                &health_cap,
+            // Buffer as many events as possible. This should be bounded, as new data can't be
+            // produced in this worker until we yield to timely.
```

Comment:
Hm, I didn't understand why we need to do this and not rely on progress events. Can you elaborate on the reasoning?

Reply:
During hydration we don't get a progress event until the very end of hydration!